library/std/src/io/copy.rs

```diff
@@ -39,6 +39,12 @@ use crate::mem::MaybeUninit;
 ///     Ok(())
 /// }
 /// ```
+///
+/// # Platform-specific behavior
+///
+/// On Linux (including Android), this function uses `copy_file_range(2)`,
+/// `sendfile(2)` or `splice(2)` syscalls to move data directly between file
+/// descriptors if possible.
```

Comment:
The `io::copy` optimizations rely on specialization, currently we're usually not documenting those since specialization itself isn't stable and so may change in the future. At a minimum there should be a caveat in there that it's only current/not guaranteed behavior.

Reply:
I see. And even beyond Rust not necessarily committing to this there are a bunch of very complicated conditions around whether `copy_file_range` will actually work or whether it needs a fallback.

I was trying here to follow the direction of <https://doc.rust-lang.org/std/io/index.html#platform-specific-behavior>: 

> Many I/O functions throughout the standard library are documented to indicate what various library or syscalls they are delegated to. This is done to help applications both understand what’s happening under the hood as well as investigate any possibly unclear semantics. Note, however, that this is informative, not a binding contract. The implementation of many of these functions are subject to change over time and may call fewer or more syscalls/library functions.

In particular, I was looking at the documentation and then the source to understand if the stdlib would even _attempt_ to use the handoff syscalls or whether there would be any benefit to trying it myself.

How about if I add to the comments

> The use of these syscalls is not guaranteed and may change in future releases.

?

Reply:
There actually was already a caveat just below this line 😄 

I have added one in the other file.